### PR TITLE
K8SPSMDB-276: Add backupSource to restore

### PR DESCRIPTION
**Pull request by egegunes**
[K8SPSMDB-276](https://jira.percona.com/browse/K8SPSMDB-276) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| scheduled-backup | passed |
| arbiter | passed |
| scaling | passed |
| operator-self-healing | passed |
| demand-backup-sharded | passed |
| service-per-pod | passed |
| security-context | passed |
| liveness | passed |
| rs-shard-migration | passed |
| smart-update | passed |
| version-service | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| users | passed |
| data-sharded | passed |
| pitr | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/640/commits/ff05871bd028937339f6bac0b326b8de096aaa1d
image: `perconalab/percona-server-mongodb-operator:PR-640-ff05871b`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| upgrade | passed |
| users | passed |
| upgrade-sharded | passed |
| pitr | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/640/commits/84acd70dc2fc39d1eb204f87a417bd86248a50a9
image: `perconalab/percona-server-mongodb-operator:PR-640-84acd70d`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| liveness | passed |
| upgrade | passed |
| smart-update | passed |
| upgrade-sharded | passed |
| version-service | passed |
| pitr | failed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/640/commits/3783c34f2973efd70c3e35ef27996afa4752a7e7
image: `perconalab/percona-server-mongodb-operator:PR-640-3783c34f`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | passed |
| one-pod | passed |
| storage | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| arbiter | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| pitr | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/640/commits/3783c34f2973efd70c3e35ef27996afa4752a7e7
image: `perconalab/percona-server-mongodb-operator:PR-640-3783c34f`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | failed |
| upgrade-consistency | passed |
| init-deploy | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| monitoring-2-0 | passed |
| scheduled-backup | passed |
| scaling | passed |
| arbiter | passed |
| operator-self-healing | passed |
| security-context | passed |
| demand-backup-sharded | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| upgrade | passed |
| version-service | passed |
| upgrade-sharded | passed |
| pitr | passed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/640/commits/2d7525b75ff10f2ad30bdb342d7e869f50c70cda
image: `perconalab/percona-server-mongodb-operator:PR-640-2d7525b7`

**Review by qJkee — APPROVED** [review]

File: pkg/apis/psmdb/v1/perconaservermongodbrestore_types.go [hunk 1]
```diff
@@ -16,12 +16,13 @@ import (
...
-	BackupName  string          `json:"backupName,omitempty"`
-	Destination string          `json:"destination,omitempty"`
-	StorageName string          `json:"storageName,omitempty"`
-	PITR        *PITRestoreSpec `json:"pitr,omitempty"`
+	ClusterName  string                            `json:"clusterName,omitempty"`
+	Replset      string                            `json:"replset,omitempty"`
+	BackupName   string                            `json:"backupName,omitempty"`
+	BackupSource *PerconaServerMongoDBBackupStatus `json:"backupSource"`
+	Destination  string                            `json:"destination,omitempty"`
+	StorageName  string                            `json:"storageName,omitempty"`
```
Comment (delgod) [inline pkg/apis/psmdb/v1/perconaservermongodbrestore_types.go:24]: please delete Destination and StorageName if they not needed

File: e2e-tests/pitr/run [hunk 2]
```diff
@@ -37,23 +52,36 @@ check_recovery() {
...
-		| $sed -e "s/type:/type: $restore_type/" \
-		| $sed -e "s/date:/date: $restore_date/" \
-		| kubectl_bin apply -f -
+	if [ -z "$backupSource" ]; then
+		cat $test_dir/conf/restore.yml \
+			| $sed -e "s/name:/name: restore-$backup_name/" \
+			| $sed -e "s/backupName:/backupName: $backup_name/" \
+			| $sed -e "s/type:/type: $restore_type/" \
+			| $sed -e "s/date:/date: $restore_date/" \
+			| $sed -e "11,13d" \
```
Comment (delgod) [inline e2e-tests/pitr/run:73]: let's be more readable here
```suggestion
			| $sed -e "/backupSource/,+2d" \
```

File: pkg/apis/psmdb/v1/perconaservermongodbrestore_types.go [hunk 3]
```diff
@@ -73,8 +74,22 @@ func (r *PerconaServerMongoDBRestore) CheckFields() error {
 		return fmt.Errorf("spec clusterName field is empty")
 	}
 
-	if len(r.Spec.BackupName) == 0 && (len(r.Spec.StorageName) == 0 || len(r.Spec.Destination) == 0) {
-		return errors.New("fields backupName or storageName and destination is empty")
+	if len(r.Spec.BackupName) == 0 && r.Spec.BackupSource == nil {
+		return errors.New("one of backupName or backupSource is required")
```
Comment (delgod) [inline pkg/apis/psmdb/v1/perconaservermongodbrestore_types.go:78]: now we have both - r.Spec.StorageName and r.Spec.BackupSource.StorageName
looks like a mistake

File: pkg/apis/psmdb/v1/perconaservermongodbrestore_types.go [hunk 4]
```diff
@@ -73,8 +74,22 @@ func (r *PerconaServerMongoDBRestore) CheckFields() error {
...
+	if r.Spec.BackupSource != nil {
+		if len(r.Spec.BackupSource.Destination) == 0 {
+			return errors.New("backupSource destination is required")
+		}
+
+		if r.Spec.BackupSource.Destination[:5] != "s3://" {
+			return errors.New("backupSource destination should use s3 protocol format")
+		}
+
+		if len(r.Spec.BackupSource.StorageName) == 0 && r.Spec.BackupSource.S3 == nil {
```
Comment (delgod) [inline pkg/apis/psmdb/v1/perconaservermongodbrestore_types.go:90]: should we also check if StorageNmae exists in the list of storages in Cluster CR?

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/apis/psmdb/v1/perconaservermongodbrestore_types.go (hunk 4 above)
Comment (egegunes) [reply pkg/apis/psmdb/v1/perconaservermongodbrestore_types.go:90]: we check it in the `r.getStorage`: https://github.com/percona/percona-server-mongodb-operator/blob/2d7525b75ff10f2ad30bdb342d7e869f50c70cda/pkg/controller/perconaservermongodbrestore/perconaservermongodbrestore_controller.go#L327

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | failed |
| init-deploy | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| monitoring-2-0 | passed |
| scaling | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| arbiter | passed |
| demand-backup-sharded | failed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| data-sharded | passed |
| pitr | passed |
| rs-shard-migration | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/640/commits/8326e7a190c2daa973e57f65dd0ee91c317fe171
image: `perconalab/percona-server-mongodb-operator:PR-640-8326e7a1`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| self-healing | passed |
| init-deploy | passed |
| operator-self-healing | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | failed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| pitr | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/640/commits/8326e7a190c2daa973e57f65dd0ee91c317fe171
image: `perconalab/percona-server-mongodb-operator:PR-640-8326e7a1`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| init-deploy | passed |
| storage | passed |
| version-service | passed |
| limits | passed |
| self-healing | passed |
| users | passed |
| scaling | passed |
| operator-self-healing | passed |
| data-sharded | passed |
| security-context | passed |
| rs-shard-migration | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| pitr | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/640/commits/8326e7a190c2daa973e57f65dd0ee91c317fe171
image: `perconalab/percona-server-mongodb-operator:PR-640-8326e7a1`

File: pkg/apis/psmdb/v1/perconaservermongodbrestore_types.go (hunk 3 above)
Comment (egegunes) [reply pkg/apis/psmdb/v1/perconaservermongodbrestore_types.go:77]: fixed. now we're using ` r.Spec.StorageName` everywhere

File: e2e-tests/pitr/run (hunk 2 above)
Comment (egegunes) [reply e2e-tests/pitr/run]: applied your suggestion

File: pkg/apis/psmdb/v1/perconaservermongodbrestore_types.go (hunk 1 above)
Comment (egegunes) [reply pkg/apis/psmdb/v1/perconaservermongodbrestore_types.go]: deleted `Destination`

**Review by qJkee — APPROVED** [review]

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| liveness | passed |
| demand-backup-sharded | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| upgrade | failed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/640/commits/f7ef0be143243d3a666ac888f8030aaa44771b16
image: `perconalab/percona-server-mongodb-operator:PR-640-f7ef0be1`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| liveness | passed |
| smart-update | passed |
| upgrade-sharded | passed |
| version-service | passed |
| pitr | passed |
| users | passed |
| data-sharded | passed |
| pitr-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/640/commits/28b6f3099c255caefd5ef39ea7065b4cb3d78dd2
image: `perconalab/percona-server-mongodb-operator:PR-640-28b6f309`

**Review by qJkee — APPROVED** [review]

File: e2e-tests/pitr/run [hunk 5]
```diff
@@ -27,14 +27,23 @@ write_initial_data() {
...
 	local name=$1
 	local idx=$2
 
 	desc "run backup $name-$idx"
-	cat $test_dir/conf/$name.yml \
-		| $sed -e "s/name:/name: $name-$idx/" \
-		| kubectl_bin apply -f -
+	cat $test_dir/conf/$name.yml |
+		$sed -e "s/name:/name: $name-$idx/" |
+		kubectl_bin apply -f -
```
[the remark this answers is not in the document]
Comment (delgod) [reply e2e-tests/pitr/run:46]: @egegunes can you please follow this advice?
reason: if you have a large multiline pipeline of very long strings >256 chars per line. it is very easy to lose `|` pipe char and the reader can start reading newline as an absolutely new command and haven't realized that this command just part of a multiline pipeline.

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| demand-backup-sharded | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| upgrade | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/640/commits/cb9855929c0a1526ca6aea9ed711f08bf20ed18f
image: `perconalab/percona-server-mongodb-operator:PR-640-cb985592`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| operator-self-healing | passed |
| smart-update | passed |
| upgrade | passed |
| version-service | passed |
| upgrade-sharded | passed |
| pitr | passed |
| users | passed |
| data-sharded | passed |
| pitr-sharded | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/640/commits/34f453a3bb8977ebeb4f2634d8c0b57bcf169c45
image: `perconalab/percona-server-mongodb-operator:PR-640-34f453a3`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| upgrade | passed |
| version-service | passed |
| upgrade-sharded | passed |
| users | passed |
| pitr | passed |
| data-sharded | passed |
| pitr-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/640/commits/d86503f2c6592f73d6a7f086f54a0e685b57b93a
image: `perconalab/percona-server-mongodb-operator:PR-640-d86503f2`

**Review by delgod — APPROVED** [review]

**Pull request merged**